doc/usage.rst

```diff
@@ -0,0 +1,241 @@
+keep track of all visited array type definitions it sees.
+The :py:meth:`~hdlparse.vhdl_parser.VhdlExtractor.is_array` method lets you query the internal list to check if a type
+is for an array.
+All IEEE standard array types are supported by default.
+Any subtypes derived from an array type will also be considered as arrays.
+
+.. code-block:: python
+
+  import hdlparse.vhdl_parser as vhdl
+
+  vhdl_ex = vhdl.VhdlExtractor()
+
+  code = '''
+  package foobar is
+    type custom_array is array(integer range <>) of boolean;
+    subtype custom_subtype is custom_array(1 to 10);
+  end package;
+  '''
+  vhdl_comps = vhdl_ex.extract_objects(code)
+
+  # These all return true:
+  print(vhdl_ex.is_array('unsigned'))
+  print(vhdl_ex.is_array('custom_array'))
+  print(vhdl_ex.is_array('custom_subtype'))
+
+Parsed array data can be saved to a file with :py:meth:`~hdlparse.vhdl_parser.VhdlExtractor.save_array_types` and
+restored with :py:meth:`~hdlparse.vhdl_parser.VhdlExtractor.load_array_types`.
+This lets you parse one set of files for type definitions and use the saved info for parsing other code at a different
+time.
+
+
+Verilog
+=======
+
+The Verilog parser is only able to extract module definitions with a port and optional parameter list.
+Verilog modules are extracted using the :py:meth:`~hdlparse.verilog_parser.VerilogExtractor.extract_objects` and
+:py:meth:`~hdlparse.verilog_parser.VerilogExtractor.extract_objects_from_source` methods.
+The latter is used when you have the code in a string.
+The former when you want to read the Veirlog source from a file.
+When parsing a file, a cache of objects is maintained so you can repeatedly call
+:py:meth:`~hdlparse.verilog_parser.VerilogExtractor.extract_objects` without reparsing the file.
+
+.. code-block:: python
+
+  with open(fname, 'rt') as fh:
+    code = fh.read()
+  vlog_mods = vlog_ex.extract_objects_from_source(code)
+
+  vlog_mods = vlog_ex.extract_objects(fname)
+
+The result is a list of extracted :py:class:`~hdlparse.verilog_parser.VerilogModule` objects.
+Each instance of this class has ``name``, ``generics``, and ``ports`` attributes.
+The ``name`` attribute is the name of the module.
+The ``generics`` attribute is a list of extracted parameters and ``ports`` is a list of the ports on the module.
+
+.. code-block:: verilog
+
+  module newstyle // This is a new style module def
+  #(parameter real foo = 8, bar=1, baz=2,
+  parameter signed [7:0] zip = 100)
+  (
+    input x, x2, inout y, y2_long_output,
+    output wire [4:1] z, z2
+  );
+  endmodule
+
+Each port and generic is an instance of :py:class:`~hdlparse.verilog_parser.VerilogParameter` containing the name, mode
+(input, output, inout), and type.
+
+.. code-block:: python
+
+  import hdlparse.verilog_parser as vlog
+
+  vlog_ex = vlog.VerilogExtractor()
+  vlog_mods = vlog_ex.extract_objects_from_source('example.v')
```

Comment:
Should be `extract_objects`

Reply:
Can you please do that change in a separated commit and push it to this branch/PR?